From 9efdce40607c9260a7f1c0bdca7386e86b46aa9b Mon Sep 17 00:00:00 2001
From: Jeegar Patel <jeegar.patel@xilinx.com>
Date: Tue, 22 Sep 2020 05:04:54 -0700
Subject: [PATCH 1625/1851] drm: xlnx: fb: Fix OVERFLOW_BEFORE_WIDEN type
 warnings

commit 4bf7458ae3fafdf28e7ac7556f6d90e42afb0d44 from
https://github.com/Xilinx/linux-xlnx.git

This patch fixes OVERFLOW_BEFORE_WIDEN type coverity warnings.

Potentially overflowing expression "mode_cmd.pitches[0] *
mode_cmd.height" with type "unsigned int" (32 bits, unsigned) is
evaluated using 32-bit arithmetic, and then used in a context that
expects an expression of type "size_t" (64 bits, unsigned).

To avoid overflow, cast either "mode_cmd.pitches[0]" or "mode_cmd.height"
to type "size_t".

Signed-off-by: Jeegar Patel <jeegar.patel@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/xlnx_fb.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/xlnx/xlnx_fb.c b/drivers/gpu/drm/xlnx/xlnx_fb.c
index e5d86bfd2280..ba8afc14381f 100644
--- a/drivers/gpu/drm/xlnx/xlnx_fb.c
+++ b/drivers/gpu/drm/xlnx/xlnx_fb.c
@@ -157,7 +157,7 @@ xlnx_fb_gem_fbdev_fb_create(struct drm_device *drm,
 	mode_cmd.pixel_format = drm_driver_legacy_fb_format(drm,
 							    size->surface_bpp,
 							    size->surface_depth);
-	if (obj->size < mode_cmd.pitches[0] * mode_cmd.height)
+	if (obj->size < (size_t)mode_cmd.pitches[0] * mode_cmd.height)
 		return ERR_PTR(-EINVAL);
 
 	return xlnx_fb_gem_fb_alloc(drm, &mode_cmd, &obj, 1, funcs);
@@ -192,7 +192,8 @@ static int xlnx_fbdev_create(struct drm_fb_helper *fb_helper,
 
 	size->surface_height *= fbdev->vres_mult;
 	bytes_per_pixel = DIV_ROUND_UP(size->surface_bpp, 8);
-	bytes = ALIGN(size->surface_width * bytes_per_pixel, fbdev->align);
+	bytes = ALIGN((size_t)size->surface_width * bytes_per_pixel,
+		      fbdev->align);
 	bytes *= size->surface_height;
 
 	obj = drm_gem_cma_create(drm, bytes);
@@ -235,7 +236,7 @@ static int xlnx_fbdev_create(struct drm_fb_helper *fb_helper,
 	drm_fb_helper_fill_info(fbi, fb_helper, size);
 	fbi->var.yres = fb->height / fbdev->vres_mult;
 
-	offset = fbi->var.xoffset * bytes_per_pixel;
+	offset = (unsigned long)fbi->var.xoffset * bytes_per_pixel;
 	offset += fbi->var.yoffset * fb->pitches[0];
 
 	drm->mode_config.fb_base = (resource_size_t)obj->paddr;
-- 
2.31.1

